tox.ini

```diff
@@ -4,20 +4,42 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py35,py36,py37,py38,pypy3,flake8
+envlist = flake8,packaging,docs,py3.5,py3.6,py3.7,py3.8,pypy3
 
 [testenv]
 basepython =
-    py35: python3.5
-    py36: python3.6
-    py37: python3.7
-    py38: python3.8
+    py3.5: python3.5
+    py3.6: python3.6
+    py3.7: python3.7
+    py3.8: python3.8
     pypy3: pypy3
-
-commands = {envpython} setup.py test
+deps =
+    pytest
+    pytest-tldr
+whitelist_externals =
+    make
+commands =
+    make
+    pytest -vv
 
 [testenv:flake8]
 basepython = python3.8
 deps =
     flake8
 commands = flake8 {posargs}
+
+[testenv:packaging]
+basepython = python3.8
+deps =
+    wheel
+    twine
+commands =
+    {envpython} setup.py sdist bdist_wheel
+    {envpython} -m twine check dist/*
+
+[testenv:docs]
+basepython = python3.8
+deps =
+    -r{toxinidir}/docs/requirements_docs.txt
+commands =
+    {envpython} setup.py build_sphinx -W
```

Comment:
As above, it would be better to use `python` instead of `{envpython}`.

Reply:
Is there a particular reason for using `python setup.py build_sphinx` instead of `cd docs; make html`, which is what our documentation recommends?

On that topic - perhaps we should adjust that documentation to recommend `tox -e docs` as the main way for building the docs. That way you don't have to worry about manually installing the documentation build dependencies or switching to the correct directory, because tox takes care of all of that.

Reply:
I opted for `build_sphinx` because `cd docs && make` requires whitelisting `cd` and `make`. Admittedly, make is already whitelisted, but given there's a convenient entry point, I figured I might as well use it.

As for updating the docs about building docs - +1 to that.

Reply:
tox has a built-in [`changedir`](https://tox.readthedocs.io/en/latest/config.html#conf-changedir) option to run a specific environment in a different directory. `cd` probably wouldn't work as expected anyway - I think tox doesn't run all commands in the same shell.

Personally I always automatically avoid using `setup.py whatever` when there is an alternative, because the `setup.py` command always seems to call some legacy `distutils` code that does something you don't want. (For example `setup.py install` not playing nicely with pip so you should use `pip install`, and `setup.py upload` sending your password in plain text so you should use `twine upload`.) Then there's also the general move towards supporting build tools other than setuptools (with PEP 517/518, etc.), so it seems logical to avoid coupling to setuptools for no good reason, in case we ever want to switch to a different build/packaging tool.

Anyway, it probably doesn't matter much in this case - `setup.py build_sphinx` seems to be a fully supported way of calling Sphinx with no indication that it's deprecated or has odd behavior or anything. As long as we choose one way of calling Sphinx and stick to it consistently, I think either `setup.py build_sphinx` or `make` in the docs folder is okay.